### PR TITLE
perf: LSP cancellation support

### LINKED ISSUE #791: LSP cancellation support
For having better performance we need to support cancel checker support.

Takes a sample:

 * User hover a quarkus property
 * The LSP request textDocument/hover is called
 * User move the mouse, it means that he don't need the hover result
 * The LSP4IJ need to cancel the textDocument/hover to notify to the server that it must cancel the textDocument/hover process

See LSP spec at https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#cancelRequest

### PR DESCRIPTION
**Pull request by angelozerr**
perf: LSP cancellation support

Fixes #791

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerExcecutionContext.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.eclipse.lsp4j.services.LanguageServer;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class LanguageServerExcecutionContext implements CancelChecker  {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerExcecutionContext.java:17]: LanguageServerExecutionContext

Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerExcecutionContext.java]: I did somerefactoring, now I rename this class with `CancellationSupport`

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerExcecutionContext.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    @Override
+    public boolean isCanceled() {
+        return canceled;
+    }
+
+    public void cancel() {
+        if (canceled == true) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerExcecutionContext.java]: fixed in CancellationSupport

**Comment by angelozerr** [comment]
The PR is ready and all operations completion, inlayhint, etc managed cancellation support. The cancel support is interesting when  language server takes some times to give a response and a new request is called.

You should see now in the trace `Sending notification '$/cancelRequest'`

```
[Trace - 09:35:35 AM] Received notification 'textDocument/publishDiagnostics'
[Trace - 09:35:35 AM] Sending notification '$/cancelRequest'
[Trace - 09:35:35 AM] Received response 'textDocument/inlayHint - (2)' in 392ms.
[Trace - 09:35:36 AM] Sending request 'textDocument/inlayHint - (3)'.
[Trace - 09:35:36 AM] Sending notification '$/cancelRequest'
[Trace - 09:35:36 AM] Received response 'textDocument/inlayHint - (3)' in 856ms.
[Trace - 09:35:36 AM] Sending notification 'textDocument/didSave'
[Trace - 09:35:36 AM] Sending notification 'textDocument/didChange'
[Trace - 09:35:37 AM] Sending notification 'textDocument/didSave'
[Trace - 09:35:37 AM] Sending notification 'textDocument/didChange'
[Trace - 09:35:37 AM] Sending request 'textDocument/inlayHint - (4)'.
[Trace - 09:35:37 AM] Sending notification '$/cancelRequest'

```

As our language server are fast, it can be difficult to produce a cancel. 

To reproduce this cancel, I created an application.properties with 1500 lines and I type fast to open / close the completion and you should see `$/cancelRequest`. I notice too that inlay hint can be slow and sometimes it is canceled with $/cancelRequest.

**Comment by angelozerr** [comment]
Please note that when a cancelRequest occurs it will not cancel the future which is created on IJ side (IndexAwareLanguageClient). We need to do some improvement on our language server

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationSupport.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		}
+		futuresToCancel.clear();
+	}
+
+	@Override
+	public void checkCanceled() {
+		// When LSP requests are called (ex : 'textDocument/completion') the LSP
+		// response
+		// items are used to compose some UI item (ex : LSP CompletionItem are translate
+		// to Eclipse ICompletionProposal).
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationSupport.java:65]: "Eclipse ICompletionProposal" ;-)

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.intellij.lsp4ij;
+
+import org.eclipse.lsp4j.services.LanguageServer;
+
+/**
+ * Itemwhich stores the initialized LSP4j language server and the language server wrapper.
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java:16]: Item which

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class LanguageServerItem {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java:18]: we'll have to turn this into a record, once Java 17 is required

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/LSPCompletionContributor.java [hunk 6]
```diff
@@ -54,32 +54,46 @@ public void fillCompletionVariants(@NotNull CompletionParameters parameters, @No
...
+            CompletableFuture<List<LanguageServerItem>> completionLanguageServersFuture = initiateLanguageServers(project, document);
+            cancellationSupport.execute(completionLanguageServersFuture);
+            ProgressManager.checkCanceled();
+
             /*
              process the responses out of the completable loop as it may cause deadlock if user is typing
              more characters as toProposals will require as read lock that this thread already have and
              async processing is occuring on a separate thread.
              */
+            ProgressManager.checkCanceled();
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/LSPCompletionContributor.java:71]: there's already a `checkCanceled()` call right before this one

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java [hunk 7]
```diff
@@ -181,29 +185,27 @@ private int getTargetOffset(PsiElement originalElement) {
...
-                                            LOGGER.warn(e.getLocalizedMessage(), e);
-                                            return null;
-                                        } catch (InterruptedException e) {
-                                            LOGGER.warn(e.getLocalizedMessage(), e);
-                                            Thread.currentThread().interrupt();
-                                            return null;
-                                        }
-                                    }).filter(Objects::nonNull).collect(Collectors.toList()));
+                                    .map(languageServer ->
+                                            cancellationSupport.execute(
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java:204]: if it runs in async, there's a small risk that at this point cancellationSupport could be null-ed by other methods

**Review by fbricon — CHANGES_REQUESTED** [review]

**Comment by fbricon** [comment]
I opened https://github.com/ia3andy/quarkus-blast/blob/main/src/main/resources/templates/GameController/partials/grid.html, quickly hovered over the inlay hints and got that exception:

```
java.lang.ClassCastException: class com.redhat.devtools.intellij.lsp4ij.LanguageServerWrapper cannot be cast to class org.eclipse.lsp4j.services.LanguageServer (com.redhat.devtools.intellij.lsp4ij.LanguageServerWrapper and org.eclipse.lsp4j.services.LanguageServer are in unnamed module of loader com.intellij.ide.plugins.cl.PluginClassLoader @72128910)
	at com.redhat.devtools.intellij.lsp4ij.operations.inlayhint.LSPInlayHintInlayProvider.lambda$createInlayPresentation$1(LSPInlayHintInlayProvider.java:182)
	at com.intellij.codeInsight.hints.presentation.PresentationFactory$referenceOnHover$hovered$1.invoke(PresentationFactory.kt:278)
	at com.intellij.codeInsight.hints.presentation.PresentationFactory$referenceOnHover$hovered$1.invoke(PresentationFactory.kt:42)
	at com.intellij.codeInsight.hints.presentation.PresentationFactory$onClick$2.invoke(PresentationFactory.kt:251)
	at com.intellij.codeInsight.hints.presentation.PresentationFactory$onClick$2.invoke(PresentationFactory.kt:42)
	at com.intellij.codeInsight.hints.presentation.OnClickPresentation$1.onClick(OnClickPresentation.kt:17)
	at com.intellij.codeInsight.hints.presentation.OnClickPresentation.mouseClicked(OnClickPresentation.kt:21)
	at com.intellij.codeInsight.hints.presentation.DynamicDelegatePresentation.mouseClicked(DynamicDelegatePresentation.kt:53)
	at com.intellij.codeInsight.hints.presentation.StaticDelegatePresentation.mouseClicked(StaticDelegatePresentation.kt:51)
	at com.intellij.codeInsight.hints.presentation.StaticDelegatePresentation.mouseClicked(StaticDelegatePresentation.kt:51)
	at com.intellij.codeInsight.hints.presentation.SequencePresentation$mouseClicked$1.invoke(SequencePresentation.kt:71)
	at com.intellij.codeInsight.hints.presentation.SequencePresentation$mouseClicked$1.invoke(SequencePresentation.kt:16)
	at com.intellij.codeInsight.hints.presentation.SequencePresentation.handleMouse(SequencePresentation.kt:62)
	at com.intellij.codeInsight.hints.presentation.SequencePresentation.mouseClicked(SequencePresentation.kt:70)
	at com.intellij.codeInsight.hints.presentation.InsetPresentation$mouseClicked$1.invoke(InsetPresentation.kt:53)
	at com.intellij.codeInsight.hints.presentation.InsetPresentation$mouseClicked$1.invoke(InsetPresentation.kt:13)
	at com.intellij.codeInsight.hints.presentation.InsetPresentation.handleMouse(InsetPresentation.kt:48)
	at com.intellij.codeInsight.hints.presentation.InsetPresentation.mouseClicked(InsetPresentation.kt:52)
	at com.intellij.codeInsight.hints.presentation.StaticDelegatePresentation.mouseClicked(StaticDelegatePresentation.kt:51)
	at com.intellij.codeInsight.hints.presentation.StaticDelegatePresentation.mouseClicked(StaticDelegatePresentation.kt:51)
	at com.intellij.codeInsight.hints.presentation.DynamicInsetPresentation$mouseClicked$1.invoke(DynamicInsetPresentation.kt:57)
	at com.intellij.codeInsight.hints.presentation.DynamicInsetPresentation$mouseClicked$1.invoke(DynamicInsetPresentation.kt:11)
	at com.intellij.codeInsight.hints.presentation.DynamicInsetPresentation.handleMouse(DynamicInsetPresentation.kt:52)
	at com.intellij.codeInsight.hints.presentation.DynamicInsetPresentation.mouseClicked(DynamicInsetPresentation.kt:56)
	at com.intellij.codeInsight.hints.presentation.RecursivelyUpdatingRootPresentation.mouseClicked(RecursivelyUpdatingRootPresentation.kt:69)
	at com.intellij.codeInsight.hints.LinearOrderInlayRenderer.mouseClicked(LinearOrderInlayRenderer.kt:113)
	at com.intellij.codeInsight.hints.presentation.listeners.InlayEditorMouseListener$mouseClicked$1.run(InlayEditorMouseListener.kt:26)
	at com.intellij.util.SlowOperations.allowSlowOperations(SlowOperations.java:149)
	at com.intellij.codeInsight.hints.presentation.listeners.InlayEditorMouseListener.mouseClicked(InlayEditorMouseListener.kt:25)
	at com.intellij.openapi.editor.impl.event.EditorEventMulticasterImpl$2.lambda$mouseClicked$1(EditorEventMulticasterImpl.java:73)
	at com.intellij.openapi.extensions.impl.ExtensionProcessingHelper.forEachExtensionSafe(ExtensionProcessingHelper.java:34)
	at com.intellij.openapi.extensions.ExtensionPointName.forEachExtensionSafe(ExtensionPointName.java:46)
	at com.intellij.openapi.editor.impl.event.EditorEventMulticasterImpl$2.mouseClicked(EditorEventMulticasterImpl.java:73)
	at com.intellij.openapi.editor.impl.EditorImpl$MyMouseAdapter.runMouseClickedCommand(EditorImpl.java:3954)
	at com.intellij.openapi.editor.impl.EditorImpl$MyMouseAdapter.mouseReleased(EditorImpl.java:3873)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6648)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3392)
	at java.desktop/java.awt.Component.processEvent(Component.java:6413)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5022)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4854)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4575)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2802)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4854)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:781)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:730)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:754)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:752)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:751)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:898)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:743)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:439)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:791)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:438)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:604)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:436)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:881)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:484)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)

```

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationSupport.java (hunk 3 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/internal/CancellationSupport.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java (hunk 4 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/LanguageServerItem.java]: fixed

File: src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/LSPCompletionContributor.java (hunk 6 above)
Comment (angelozerr) [reply src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/completion/LSPCompletionContributor.java]: fixed

**Comment by angelozerr** [comment]
> Also saw this in my logs:

I don't know -(

**Comment by angelozerr** [comment]
> I opened https://github.com/ia3andy/quarkus-blast/blob/main/src/main/resources/templates/GameController/partials/grid.html, clicked on an inlay hint and got that exception:

Should be fixed

**Comment by fbricon** [comment]
Thanks @angelozerr

**Pull request merged**